### DISCUSSION TITLE
Data Dashboard

### DISCUSSION THREAD

**Discussion by zachberry** [ucfopen/Obojobo#1779] [Ideas]
This discussion can keep track of our overall goal of implementing a well-needed data dashboard for Obojobo. 

We see a few phases to implement this:

* Write queries to extract data we're interested in (Phase 0) - #1778 
* Author-accessible raw data exports (Phase 1) - #1454 
* Implement fetching data via background worker (Phase 2)
  * Create an optional background worker (we probably want a dumb single server fallback)
  * Create a queue of tasks for background worker
  * Update our architecture to add a background worker
* Basic data dashboard (Phase 3)
* Enhance dashboard with Chen-like insights (Phase 4)

**Comment by iturgeon** [comment]
Spitballing, this may need a new docker container that starts a worker.  Maybe a new pm2 configuration to run & monitor the background workers within the existing docker container.

A separate container would make it easier to scale the workers.

We should also be able to run the worker via the Heroku configuration (they offer a free worker process).

**Comment by zachberry** [comment]
We also want to eventually include any insights that adaptive features would provide

**Comment by zachberry** [comment]
One potential visualization idea from Dr. Chen:

![image](https://user-images.githubusercontent.com/73479/118495704-728aba80-b6f1-11eb-8202-7e0be3af9601.png)

Generated using process mining, showing one module with 2 pages and 5 attempts

**Comment by zachberry** [comment]
Some notes from Dr. Chen on what would be helpful:

A couple of quick thoughts on assessment stats:

1. For instructors it is important to be able to see assessment stats for students in a given class.
2. For MC problems instructors will want to see distribution of students selecting each option.
3. A duration distribution histogram carries very important information about students' engagement. Duration distributions are usually more helpful when shown on log scale.
4. It will be excellent if you can show the percentage of correct answers for students who spent at least 60 seconds on the assessment, for example.

**Comment by zachberry** [comment]
Notes of possible nice additions to #1864 

Overall:

* Add in help on what the fields mean (If possible with the data grid component, I'd like to add (?) icons next to the header column items)

Final Assessment Scores table:

* Add a button to jump to the All Attempt Scores view, with that users' attempts selected
* Add in number of attempts taken

Attempt Scores table:

* Add in a total time

Things to add:

* Grade distribution graph
* Attempt submit time graph
* Attempt and assessment score graph for each attempt for a user
* Summary information:
	* First record time
	* Last record time
	* List of all courses, resource links, draft ids and draft versions
